src/monattack.cpp

```diff
@@ -4720,6 +4726,142 @@ bool mattack::riotbot( monster *z )
+        target->add_msg_player_or_npc( msg_type, _( "The %s lunges at you, but you dodge!" ),
+                                       _( "The %s lunges at <npcname>, but they dodge!" ),
+                                       z->name() );
+        if( !uncanny ) {
+            target->on_dodge( z, z->type->melee_skill * 2 );
+            target->add_msg_player_or_npc( msg_type, _( "The %s lunges at you, but you dodge!" ),
+                                           _( "The %s lunges at <npcname>, but they dodge!" ),
+                                           z->name() );
+        }
+        return true;
+    }
+    const std::string old_name = z->name();
+    const bool could_see_z = g->u.sees( *z );
+    tripoint const target_pos = target->pos();
+    const std::string target_name = target->disp_name();
+    damage_instance damage( z->type->melee_damage );
+    damage.mult_damage( 1.33f );
+    int damage_dealt = target->deal_damage( z, bp_torso, damage_instance( DT_STAB, rng( 10, 20 ),
+                                            rng( 5, 15 ) ) ).total_damage();
+    if( damage_dealt > 0 ) {
+        auto msg_type = target == &g->u ? m_bad : m_warning;
+        target->add_msg_player_or_npc( msg_type,
+                                       _( "The %1$s impales yor chest for %2$d damage!" ),
+                                       _( "The %1$s impales <npcname>'s chest for %2$d damage!" ),
+                                       z->name(), damage_dealt );
+    } else {
+        target->add_msg_player_or_npc(
+            _( "The %1$s attempts to burrow itself into you, but is stopped by your armor!" ),
+            _( "The %1$s slashes at <npcname>'s %2$s, but is stopped by its glances off armor!" ),
+            z->name() );
+        return true;
+    }
+    if( target->is_dead_state() && g->is_empty( target_pos ) &&
+        target->made_of_any( Creature::cmat_flesh ) ) {
+        z->allow_upgrade();
+        z->try_upgrade( false );
+        z->setpos( target_pos );
+        const std::string upgrade_name = z->name();
+        const bool can_see_z_upgrade = g->u.sees( *z );
+        if( could_see_z && can_see_z_upgrade ) {
+            add_msg( m_warning, _( "The %1$s burrows within %2$s corpse and a %3$s emerges from the remains!" ),
+                     old_name,
+                     target_name, upgrade_name );
+        } else if( could_see_z ) {
+            add_msg( m_warning, _( "The %1$s burrows within %2$s corpse!" ), old_name, target_name );
+        } else if( can_see_z_upgrade ) {
+            add_msg( m_warning, _( "A %1$s emerges from %2$s corpse!" ), target->disp_name(), target_name );
+        }
+    }
+    return true;
+}
+
+bool mattack::leech_spawner( monster *z )
+{
+    const bool u_see = g->u.sees( *z );
+    std::list<monster *> allies;
+    for( monster &candidate : g->all_monsters() ) {
+        if( candidate.type->in_species( LEECH_PLANT ) && !candidate.type->has_flag( MF_IMMOBILE ) ) {
```

Comment:
```suggestion
        if( candidate.in_species( LEECH_PLANT ) && !candidate.has_flag( MF_IMMOBILE ) ) {
```
Those functions exist as part of the `monster` class, so you don't need to check for them within the monster type.